packages/backend-app-api/src/services/implementations/rootFeatureRegistry/InstanceRegistration.ts

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  DiscoveryService,
+  RootFeatureRegistryService,
+} from '@backstage/backend-plugin-api';
+import fetch from 'node-fetch';
+import { PluginRegistrations } from '../discovery/MultipleBackendHostDiscovery';
+
+export class InstanceRegistration {
+  #rootFeatureRegistry: RootFeatureRegistryService;
+  #discovery: DiscoveryService;
+  #gatewayUrl: string;
+  #instanceUrl: string;
+  #registrationIntervalId: NodeJS.Timeout | undefined;
+  #plugins: Record<string, { internal: string; external: string }> = {};
+
+  constructor(options: {
+    rootFeatureRegistry: RootFeatureRegistryService;
+    discovery: DiscoveryService;
+    gatewayUrl: string;
+    instanceUrl: string;
+  }) {
+    this.#discovery = options.discovery;
+    this.#rootFeatureRegistry = options.rootFeatureRegistry;
+    this.#gatewayUrl = options.gatewayUrl;
+    this.#instanceUrl = options.instanceUrl;
+  }
+
+  private async submitRegistration(plugins: PluginRegistrations) {
+    try {
+      const response = await fetch(
+        `${this.#gatewayUrl}/api/discovery/install`,
+        {
+          method: 'POST',
+          body: JSON.stringify({ instanceUrl: this.#instanceUrl, plugins }),
+          headers: {
+            'Content-Type': 'application/json',
+          },
+        },
+      );
+      if (response.ok) {
+        return true;
+      }
+    } catch (err) {
+      // skip network errors.
+    }
+    return false;
+  }
+
+  async register() {
+    this.#registrationIntervalId = this.intervalWithImmediateExecution(
+      async () => {
+        const registered = await this.submitRegistration(this.#plugins);
+        console.log('registered', registered, this.#plugins);
+        if (registered) {
+          clearInterval(this.#registrationIntervalId);
+          this.#registrationIntervalId = undefined;
+        }
+        // The interval numbers will _definitely_ need tweaking.
+      },
+      10000,
+    );
+  }
+
+  private async heartbeat() {
+    try {
+      const response = await fetch(`${this.#gatewayUrl}/api/discovery/check`, {
```

Comment:
Dependency ordering means we can't use `discoveryService` here. Hard-coded/config-defined doesn't seem that bad.